Assets/MixedRealityToolkit/InputModule/Scripts/Utilities/Interactions/TwoHandManipulatable.cs

```diff
@@ -0,0 +1,631 @@
+
+            m_handRefDistance = Vector3.Distance(newHandPosition, pivotPosition);
+            m_objRefDistance = Vector3.Distance(manipulationRoot.position, pivotPosition);
+
+            var objDirectoin = Vector3.Normalize(manipulationRoot.position - pivotPosition);
+            var handDirection = Vector3.Normalize(newHandPosition - pivotPosition);
+
+            // We transform the forward vector of the object, the direction of the object, and the direction of the hand
+            // to camera space so everything is relative to the user's perspective.
+            objDirectoin = Camera.main.transform.InverseTransformDirection(objDirectoin);
+            handDirection = Camera.main.transform.InverseTransformDirection(handDirection);
+
+            // Store the original rotation between the hand an object
+            m_gazeAngularOffset = Quaternion.FromToRotation(handDirection, objDirectoin);
+            m_draggingPosition = manipulationRoot.position;
+        }
+
+        public Vector3 Update(Vector3 centroid, Vector3 manipulationObjectPosition)
+        {
+            var newHandPosition = centroid;
+            var pivotPosition = GetHandPivotPosition();
+
+            // Compute the pivot -> hand vector in camera space
+            var newHandDirection = Vector3.Normalize(newHandPosition - pivotPosition);
+            newHandDirection = Camera.main.transform.InverseTransformDirection(newHandDirection);
+
+            // The direction the object should face is the current hand direction rotated by the original hand -> object rotation.
+            var targetDirection = Vector3.Normalize(m_gazeAngularOffset * newHandDirection);
+            targetDirection = Camera.main.transform.TransformDirection(targetDirection);
+
+            // Compute how far away the object should be based on the ratio of the current to original hand distance
+            var currentHandDistance = Vector3.Magnitude(newHandPosition - pivotPosition);
+            var distanceRatio = currentHandDistance / m_handRefDistance;
+            var distanceOffset = distanceRatio > 0 ? (distanceRatio - 1f) * DistanceScale : 0;
+            var targetDistance = m_objRefDistance + distanceOffset;
+
+            var newPosition = pivotPosition + (targetDirection * targetDistance);
+
+            var newDistance = Vector3.Distance(newPosition, pivotPosition);
+            if (newDistance > 4f)
+            {
+                newPosition = pivotPosition + Vector3.Normalize(newPosition - pivotPosition) * 4f;
+            }
+
+            m_draggingPosition = newPosition;
+
+
+            return m_draggingPosition;
+        }
+
+        /// <returns>A point that is below and just in front of the head.</returns>
+        public static Vector3 GetHandPivotPosition()
+        {
+            Vector3 pivot = Camera.main.transform.position + new Vector3(0, -0.2f, 0) - Camera.main.transform.forward * 0.2f; // a bit lower and behind
```

Comment:
use `CameraCache.Main`

Reply:
Out of curiosity, why is using `CameraCache.Main` better than Camera.main?

Reply:
Cause tag lookups are expensive and this eliminates the issue.